File: Ruby/9_linked_list/21_Merge_Two_Sorted_Lists.rb
class ListNode
  attr_accessor :val, :next

  def initialize(val = 0, _next = nil)
    @val = val
    @next = _next
  end

  def to_s
    node = self
    result = "["
    while node != nil
      result += node.val.to_s
      result += " -> " if node.next != nil
      node = node.next
    end
    result + "]"
  end
end

def merge_two_lists (list1, list2)
  return list1 if list2.nil?
  return list2 if list1.nil?

  if list1.val < list2.val
    list1.next = merge_two_lists(list1.next, list2)
    list1
  else
    list2.next = merge_two_lists(list1, list2.next)
    list2
  end
end

def test_merge_two_lists(list1, list2)
  puts "\nLista 1: #{list1.to_s}"
  puts "Lista 2: #{list2.to_s}"

  start = Time.now
  result = merge_two_lists(list1, list2)
  finish = Time.now

  puts "Resultado: #{result.to_s}"
  puts "Tempo de execução: #{((finish - start) * 1_000_000_000).to_i}ns\n"
end

# Criando listas para teste
list1 = ListNode.new(1, ListNode.new(2, ListNode.new(4)))
list2 = ListNode.new(1, ListNode.new(3, ListNode.new(4)))
test_merge_two_lists(list1, list2)

list3 = ListNode.new
list4 = ListNode.new
test_merge_two_lists(list3, list4)

list5 = ListNode.new
list6 = ListNode.new(0)
test_merge_two_lists(list5, list6)

File: Go/9_linked_list/21_Merge_Two_Sorted_Lists.go
package main

import (
	"fmt"
	"time"
)

type ListNode struct {
	Val  int
	Next *ListNode
}

func (l *ListNode) String() string {
	if l == nil {
		return "[]"
	}
	result := ""
	for l != nil {
		result += fmt.Sprintf("%d -> ", l.Val)
		l = l.Next
	}
	result = result[:len(result)-4] // Remove o último ' -> '
	return "[" + result + "]"
}

func mergeTwoLists(l1 *ListNode, l2 *ListNode) *ListNode {
	preheat := &ListNode{}
	prev := preheat

	for l1 != nil && l2 != nil {
		if l1.Val <= l2.Val {
			prev.Next = l1
			l1 = l1.Next
		} else {
			prev.Next = l2
			l2 = l2.Next
		}
		prev = prev.Next
	}

	if l1 != nil {
		prev.Next = l1
	} else {
		prev.Next = l2
	}

	return preheat.Next
}

func testMergeTwoLists(list1 *ListNode, list2 *ListNode) {
	fmt.Println("List 1: ", list1)
	fmt.Println("List 2: ", list2)

	start := time.Now()
	result := mergeTwoLists(list1, list2)
	elapsed := time.Since(start)

	fmt.Println("Result: ", result)
	fmt.Println("Execution time: ", elapsed)
}

func main() {

	list1 := &ListNode{1, &ListNode{2, &ListNode{4, nil}}}
	list2 := &ListNode{1, &ListNode{3, &ListNode{4, nil}}}
	testMergeTwoLists(list1, list2)

	list3 := &ListNode{}
	list4 := &ListNode{}
	testMergeTwoLists(list3, list4)

	list5 := &ListNode{}
	list6 := &ListNode{0, nil}
	testMergeTwoLists(list5, list6)
}
